go/config/config.go

```diff
@@ -0,0 +1,316 @@
+	UnknownKeyInDefault ConfigParseErrorType = "unknown key in default section"
+
+	// UnknownKeyInService indicates that the configuration key in the service section is unknown.
+	UnknownKeyInService ConfigParseErrorType = "unknown key in 'service' section"
+
+	// InvalidValueForKey indicates that parsing the configuration value failed.
+	InvalidValueForKey ConfigParseErrorType = "invalid value for key"
+)
+
+// ConfigParseError represents an error that happened while parsing a GLOME configuration.
+type ConfigParseError struct {
+	LineNum     int
+	ErrorType   ConfigParseErrorType
+	Description string
+}
+
+// Error satisfies the Go `error` interface.
+func (e ConfigParseError) Error() string {
+	descriptionSeparator := ""
+	if e.Description != "" {
+		descriptionSeparator = ": "
+	}
+	return fmt.Sprintf("config file parsing failed in line %d (%s%s%s)", e.LineNum, e.ErrorType, descriptionSeparator, e.Description)
+}
+
+var (
+	sectionAssigners = map[string]func(cfg *Config, lineNum int, key, value string) error{
+		"default": assignDefaultSection,
+		"service": assignServiceSection,
+	}
+)
+
+// Parse parses a GLOME ini-style configuration file to a Config struct.
+func Parse(r io.Reader) (*Config, error) {
+	s := bufio.NewScanner(r)
+	currentSection := "default"
+	lineNum := 0
+	cfg := new(Config)
+	for s.Scan() {
+		lineNum++
+		txt := strings.TrimSpace(s.Text())
+		switch {
+		case len(txt) == 0, txt[0] == '#', txt[0] == ';':
+			// Purely whitespace, or a comment.
+			continue
+		case txt[0] == '[':
+			// Section header
+			end := strings.IndexByte(txt, ']')
+			if end == -1 {
+				return nil, ConfigParseError{lineNum, BadSectionName, "couldn't find closing ]"}
+			}
+			currentSection = txt[1:end]
+			if len(currentSection) == 0 {
+				return nil, ConfigParseError{lineNum, BadSectionName, "section name was empty"}
+			}
+			if _, ok := sectionAssigners[currentSection]; !ok {
+				return nil, ConfigParseError{lineNum, UnknownSection, currentSection}
+			}
+		default:
+			// Key value config option.
+			key, value, err := parseKeyValue(txt)
+			if err != nil {
+				return nil, ConfigParseError{lineNum, BadKeyValue, err.Error()}
+			}
+
+			assignValue, ok := sectionAssigners[currentSection]
+			if !ok {
+				// We shouldn't end up here since we validate section names as we assign them.
+				// However, just in case...
+				return nil, ConfigParseError{lineNum, UnknownSection, currentSection}
+			}
+			if err := assignValue(cfg, lineNum, key, value); err != nil {
+				return nil, err
+			}
+		}
+
+	}
+	return cfg, nil
+}
+
+func assignDefaultSection(cfg *Config, lineNum int, key, value string) error {
+	var err error
+	switch key {
+	case "auth-delay":
+		err = interpretPositiveInt(value, &cfg.AuthDelay)
+	case "input-timeout":
+		err = interpretPositiveInt(value, &cfg.InputTimeout)
+	case "config-path":
+		cfg.ConfigPath = value
+	case "ephemeral-key":
+		err = interpretPrivateKey(value, hex.DecodeString, &cfg.EphemeralKey)
```

Comment:
Can we yell loudly when this is set?

Reply:
Now we require an AllowInsecureOptions parameter.